File: public/code/trendstrength__dpo.py
from common import make_signal, ema, sma, rsi, macd, atr, bollinger_bands, vwap, swing_highs, swing_lows
import math


def _sign(x):
    return (x > 0) - (x < 0)


def mk(c, i, side, a, conf, reason, m=2):
    cur = c[i]
    if side == "long":
        sl = cur['close'] - m * a[i]
        r = cur['close'] - sl
        return make_signal(signal="long", entry=cur['close'], stop_loss=sl, take_profit=[cur['close'] + r * 1.5, cur['close'] + r * 2.5, cur['close'] + r * 4], confidence=conf, reason=reason)
    sl = cur['close'] + m * a[i]
    r = sl - cur['close']
    return make_signal(signal="short", entry=cur['close'], stop_loss=sl, take_profit=[cur['close'] - r * 1.5, cur['close'] - r * 2.5, cur['close'] - r * 4], confidence=conf, reason=reason)
def aroonCalc(c, end, p):
    win = c[end - p: end + 1]
    hiIdx = 0
    loIdx = 0
    for k in range(0, len(win)):
        if win[k]['high'] >= win[hiIdx]['high']:
            hiIdx = k
        if win[k]['low'] <= win[loIdx]['low']:
            loIdx = k
    up = ((p - (p - hiIdx)) / p) * 100
    dn = ((p - (p - loIdx)) / p) * 100
    return {'up': up, 'dn': dn}
def dpo(c):
    if len(c) < 40:
        return make_signal(reason="Insufficient data")
    closes = [x['close'] for x in c]
    p = 20
    shift = math.floor(p / 2) + 1
    ma = sma(closes, p)
    i = len(c) - 1
    a = atr(c, 14)
    dpoNow = closes[i - shift] - ma[i]
    dpoPrev = closes[i - 1 - shift] - ma[i - 1]
    if dpoPrev <= 0  and  dpoNow > 0:
        return mk(c, i, "long", a, 0.67, "DPO crossed above zero")
    if dpoPrev >= 0  and  dpoNow < 0:
        return mk(c, i, "short", a, 0.67, "DPO crossed below zero")
    return make_signal(reason=f"DPO {dpoNow}")

File: public/code/trendstrength__dpo.ts
import { Candle, Signal, makeSignal, sma, ema, atr } from "../indicators";

const mk = (c: Candle[], i: number, side: "long" | "short", a: number[], conf: number, reason: string, m = 2): Signal => {
  const cur = c[i];
  if (side === "long") { const sl = cur.close - m * a[i], r = cur.close - sl; return makeSignal({ signal: "long", entry: cur.close, stop_loss: sl, take_profit: [cur.close + r * 1.5, cur.close + r * 2.5, cur.close + r * 4], confidence: conf, reason }); }
  const sl = cur.close + m * a[i], r = sl - cur.close; return makeSignal({ signal: "short", entry: cur.close, stop_loss: sl, take_profit: [cur.close - r * 1.5, cur.close - r * 2.5, cur.close - r * 4], confidence: conf, reason });
};
function aroonCalc(c: Candle[], end: number, p: number) {
  const win = c.slice(end - p, end + 1);
  let hiIdx = 0, loIdx = 0;
  for (let k = 0; k < win.length; k++) { if (win[k].high >= win[hiIdx].high) hiIdx = k; if (win[k].low <= win[loIdx].low) loIdx = k; }
  const up = ((p - (p - hiIdx)) / p) * 100, dn = ((p - (p - loIdx)) / p) * 100;
  return { up, dn };
}

export function dpo(c: Candle[]): Signal {
  if (c.length < 40) return makeSignal({ reason: "Insufficient data" });
  const closes = c.map((x) => x.close), p = 20, shift = Math.floor(p / 2) + 1, ma = sma(closes, p), i = c.length - 1, a = atr(c, 14);
  const dpoNow = closes[i - shift] - ma[i], dpoPrev = closes[i - 1 - shift] - ma[i - 1];
  if (dpoPrev <= 0 && dpoNow > 0) return mk(c, i, "long", a, 0.67, "DPO crossed above zero");
  if (dpoPrev >= 0 && dpoNow < 0) return mk(c, i, "short", a, 0.67, "DPO crossed below zero");
  return makeSignal({ reason: `DPO ${dpoNow.toFixed(2)}` });
}
